### ISSUE TITLE
Write context with complex expressions should not be allowed

### ISSUE THREAD

**Issue by initBasti** [GothenburgBitFactory/taskwarrior#2549]
* What command(s) did you run?
```bash
basti@basti-TUXEDO-Book-XA1510:~$ task add pro:Work.Administration pri:H Test task ABC +work +test
Created task 45.
The project 'Work.Administration' has changed.  Project 'Work.Administration' is 99% complete (2 of 238 tasks remaining).

basti@basti-TUXEDO-Book-XA1510:~$ task 45
No command specified - assuming 'information'.

Name          Value                                                      
ID            45                                                         
Description   or Test task ABC                                           
Status        Pending                                                    
Project       Work.Administration
Entered       2021-07-31 08:45:32 (8s)                                   
Last modified 2021-07-31 08:45:32 (8s)
Tags          work freelance test                                        
Virtual tags  LATEST PENDING PRIORITY PROJECT READY TAGGED UDA UNBLOCKED
UUID          4804e2a7-875d-4cbd-913c-6135773bc1c7                       
Urgency          8
Priority      H                                                          

    project             1 *    1 =      1
    tags                1 *    1 =      1
    UDA priority.H      1 *    6 =      6
                                   ------
                                        8

Filter: ( ( ( id == 45 ) ) )
```
* What did you expect to happen?
Creation of a task with exactly the given description and tags.
* What actually happened?
'or' was prepended and the `freelance` tag added
* Paste the output of the ```task diag``` command.
```bash
basti@basti-TUXEDO-Book-XA1510:~$ task diag

task 2.6.0
   Platform: Linux

Compiler
    Version: 9.3.0
       Caps: +stdc +stdc_hosted +LP64 +c8 +i32 +l64 +vp64 +time_t64
 Compliance: C++17

Build Features
     Commit: 607baa081
      CMake: 3.16.3
    libuuid: libuuid + uuid_unparse_lower
  libgnutls: 3.6.13
 Build type: 

Configuration
       File: /home/basti/.taskrc (found), 2311 bytes, mode 100755
       Data: /home/basti/.task (found), dir, mode 40755
    Locking: Enabled
         GC: Enabled
     Server: taskwarrior.inthe.am:53589
         CA: /home/basti/.task/certs/ca.cert.pem, readable, 2405 bytes
Certificate: /home/basti/.task/certs/private.certificate.pem, readable, 2650 bytes
        Key: /home/basti/.task/certs/private.key.pem, readable, 20132 bytes
      Trust: ignore hostname
    Ciphers: NORMAL
      Creds: inthe_am/basti1992md/************************************

Hooks
     System: Enabled
   Location: /home/basti/.task/hooks
     Active: on-modify.timewarrior (executable)
   Inactive: 

Tests
   Terminal: 323x70
       Dups: Scanned 1681 tasks for duplicate UUIDs:
             No duplicates found
 Broken ref: Scanned 1681 tasks for broken references:
             No broken references found
```

I recently switched from the stable version to the latest development version, so I expected to run into some trouble, but I would like to find a way to fix this.

I have three contexts:
```bash
Context 'work' with 
* read filter: '+work or +freelance'
* write filter: '+work or +freelance'

Context 'study' with 
* read filter: '+study or +programming and -work or +coding or +hacking'
* write filter: '+study or +programming and -work or +coding or +hacking'

Context 'freetime' with 
* read filter: '+home or +free or +relax'
* write filter: '+home or +free or +relax'
```
And it looks like the `or +freelance` leaks into the task creation.

Here is another example within the study context:
```bash
basti@basti-TUXEDO-Book-XA1510:~$ task context study
Context 'study' set. Use 'task context none' to remove.
basti@basti-TUXEDO-Book-XA1510:~$ task add pro:Study pri:H Test task study ABC +study +test
Created task 45.
The project 'Study' has changed.  Project 'Study' is 85% complete (6 of 40 tasks remaining).

basti@basti-TUXEDO-Book-XA1510:~$ task 45 information

Name          Value                                                      
ID            45                                                         
Description   or and -work or or Test task study ABC                     
Status        Pending                                                    
Project       Study
Entered       2021-07-31 08:55:37 (9s)                                   
Last modified 2021-07-31 08:55:37 (9s)
Tags          study programming coding hacking test                      
Virtual tags  LATEST PENDING PRIORITY PROJECT READY TAGGED UDA UNBLOCKED
UUID          15ed7cba-6486-4824-b447-4be5adcb6ad6                       
Urgency          8
Priority      H                                                          

    project             1 *    1 =      1
    tags                1 *    1 =      1
    UDA priority.H      1 *    6 =      6
                                   ------
                                        8

Filter: ( ( ( id == 45 ) ) )
```

Greetings,
Sebastian

**Comment by tbabej** [comment]
Hi @initBasti,

the crux of the problem is the fact that the `write` context is defined as an expression with `'+work or +freelance'`. While this is fine for `read` context, which is added as a filter for displaying tasks in the reports, the `write` context is used to add new tasks within the context (see #2402 if you're interested in the implementation details!).

In order to resolve this issue, the user needs to decide what should happen with newly added tasks while the context `work` is active. There are basically three options:
* you don't want new tasks to be created with any context-related tags: set `task config context.work.write ''`
* you want new tasks to be created with tag +work: set `task config context.work.write '+work'`
* you want new tasks to be created with tag +freelance: set `task config context.work.write '+freelance'`

You can see the three options demostrated here (from the session where I was checking this :slightly_smiling_face: )
```bash
[warrior@tw-2.6.0 ~]$ task context define work '+work or +freelance'
The filter '+work or +freelance' matches 0 pending tasks. Do you wish to continue? (yes/no) yes
Are you sure you want to add 'context.work.read' with a value of '+work or +freelance'? (yes/no) yes
Are you sure you want to add 'context.work.write' with a value of '+work or +freelance'? (yes/no) yes
Context 'work' defined (read, write). Use 'task context work' to activate.
[warrior@tw-2.6.0 ~]$ task context work
Context 'work' set. Use 'task context none' to remove.

# Task that reproduces the problem you report
[warrior@tw-2.6.0 ~]$ task add test
Created task 1.

# No context modifications applied, the task will be outside of the context unless manually specified (this is pre-2.6.0 behaviour)
[warrior@tw-2.6.0 ~]$ task config context.work.write ''
Are you sure you want to change the value of 'context.work.write' from '+work or +freelance' to ''? (yes/no) yes
Config file /home/warrior/.taskrc modified.
[warrior@tw-2.6.0 ~]$ task add no modifications to added tasks
Created task 2.

# Just +work added
[warrior@tw-2.6.0 ~]$ task config context.work.write '+work'
Are you sure you want to change the value of 'context.work.write' from '' to '+work'? (yes/no) yes
Config file /home/warrior/.taskrc modified.
[warrior@tw-2.6.0 ~]$ task add this should have tag work
Created task 3.

# Just +freelance added
[warrior@tw-2.6.0 ~]$ task config context.work.write '+freelance'
Are you sure you want to change the value of 'context.work.write' from '+work' to '+freelance'? (yes/no) yes
Config file /home/warrior/.taskrc modified.
[warrior@tw-2.6.0 ~]$ task add this should have tag freelance
Created task 4.
```

Displaying the tasks we can see this worked as expected (the no modifications task ID 2 is not even displayed, because it does not match the active context):
```
[warrior@tw-2.6.0 ~]$ task

ID Age   Tag            Description                    Urg 
 1  1min freelance work or test                         0.9
 3 26s   work           this should have tag work       0.8
 4  5s   freelance      this should have tag freelance  0.8

```

I think this still needs to be documented better before the release, or perhaps a check in the `task context define` should be added to prevent user from defining a `write` context that has logical expressions. The problem always has to be resolved by the user after all, because there no clear mapping from `+work or +freelance` expression to task modifications that we can do without consulting the user.

**Comment by initBasti** [comment]
Hey @tbabej,

Thank you for explaining that issue.
Regarding your last point:

> I think this still needs to be documented better before the release, or perhaps a check in the task context define should be added to prevent user from defining a write context that has logical expressions.

Yes I believe there should be a check as a write context with a logical expression seems to be just broken as the description is malformed and tags are added that were not explicitly chosen.

I didn't notice, that feature but now that I see it, I quite like it as I save typing '+work' or '+study' for each task. It is just a little unintuitive when switching to the new version as it is normally expected that whatever I type in is what I get. Wouldn't it be better to set the write context empty by default and the user has to explicitly choose the context so that results are as expected? Or is that difficult for existing configurations?

Oh and also I notice that this doesn't work when I provide other tags:
```bash
basti@basti-TUXEDO-Book-XA1510:~/Documents/Programming_Projects/PyProjects/plenty_rait_tool$ task add pro:Work.Administration pri:H test 3 +test
Created task 44.
The project 'Work.Administration' has changed.  Project 'Work.Administration' is 99% complete (2 of 238 tasks remaining).
basti@basti-TUXEDO-Book-XA1510:~/Documents/Programming_Projects/PyProjects/plenty_rait_tool$ task 44 information

Name          Value                                                      
ID            44                                                         
Description   test 3                                                     
Status        Pending                                                    
Project       Work.Administration
Entered       2021-07-31 19:28:47 (4s)                                   
Last modified 2021-07-31 19:28:47 (4s)
Tags          test                                                       
Virtual tags  LATEST PENDING PRIORITY PROJECT READY TAGGED UDA UNBLOCKED
UUID          a30bfa3c-acff-4a47-a9eb-8818548ec323                       
Urgency        7.8
Priority      H                                                          

    project             1 *    1 =      1
    tags              0.8 *    1 =    0.8
    UDA priority.H      1 *    6 =      6
                                   ------
                                      7.8

Filter: ( ( ( id == 44 ) ) )
basti@basti-TUXEDO-Book-XA1510:~/Documents/Programming_Projects/PyProjects/plenty_rait_tool$ task context

Name       Type  Definition                                              Active
freetime   read  +home or +free or +relax                                no    
           write +home or +free or +relax                                no
open-study read  +free or +study or +programming or +coding or +hacking  no    
           write +free or +study or +programming or +coding or +hacking  no
study      read  +study or +programming and -work or +coding or +hacking no    
           write +study or +programming and -work or +coding or +hacking no
work       read  +work or +freelance                                     yes   
           write +work                                                   yes

Use 'task context none' to unset the current context.

```

Is that expected or a malfunction? I usually apply 4-5 tags for each task and I would love to skip the default ones that I just use to assign them to a context.

Thank you for the quick answer, the detailed description and the great feature :+1: :+1:

Greetings,
Sebastian

**Comment by tbabej** [comment]
> I didn't notice, that feature but now that I see it, I quite like it as I save typing '+work' or '+study' for each task. It is just a little unintuitive when switching to the new version as it is normally expected that whatever I type in is what I get. Wouldn't it be better to set the write context empty by default and the user has to explicitly choose the context so that results are as expected? Or is that difficult for existing configurations?

I think there are two cases:
* a simple context definition, that can be directly mapped to modifications. Example: `+work`
* a more complex context definition, that cannot be directly mapped (because it contains boolean expressions and other filter-related stuff). Example: `+urgent or due.before:tomorrow-8h`

What I would propose is that for a simple context definition, the write context should be populated automatically. For a complex context definition, we should prompt the user for clarification when defining the context through `task context define`.

> Oh and also I notice that this doesn't work when I provide other tags: [...]

This looks like a bug to me. I'll open a separate issue for this one.

**Comment by sebu06** [comment]
One first shot would be to check if the value of the context contains an op in CmdContext.cpp:176 like this:
```
Lexer lex (value);
    std::string lexeme;
    Lexer::Type lextype;
    bool containsOp = false;
    while (lex.token (lexeme, lextype)) {
      if(lextype == Lexer::Type::op)
        containsOp = true;
    }
 ```
the more elegant way
```
for(auto a : Context::getContext ().cli2._args ) {
      if(a._lextype == Lexer::Type::op)
        containsOp = true;
    }
```
would not work, since add filter adds parathesis to the filter and thus the filter always contains ops. The more complex, maybe better way would be to evaluate the expression and check if it can be used for write, e.g. and could also be used in a write context or `(+foo +bar)` would also be a valid expression for a write context.

The other question is what to do if that fails - one way would be to create an empty write context and tell the user to set it excplicitly using `task config context.<name>.write`. (preferred by me) The other option would be to interactively ask the user (not really the taskwarrior style, I think) and the last one would be to modify the context define command to provide both read and write commands and ask the user to rerun the command.

I think no additional check is required if the config value is set to an invalid value, since hopefully users modifying the config know what they are doing.

Regarding the other thing (adding the tags, I'm currently looking into it).

**Comment by tbabej** [comment]
I was able to take what Sebastian started here and convert it into a PR. The trick to making this work was copying the analyzed tokens before the context filter was injected for analysis. This allowed me to get the original context filter tokens and subject them to analysis :slightly_smiling_face:
